packages/js/components/src/rich-text-editor/rich-text-editor.tsx

```diff
@@ -0,0 +1,94 @@
+/**
+ * External dependencies
+ */
+import { BlockEditorProvider } from '@wordpress/block-editor';
+import { BlockInstance } from '@wordpress/blocks';
+import { SlotFillProvider } from '@wordpress/components';
+import { debounce } from 'lodash';
+import {
+	createElement,
+	useCallback,
+	useEffect,
+	useState,
+	useRef,
+} from '@wordpress/element';
+import React from 'react';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore No types for this exist yet.
+// eslint-disable-next-line @woocommerce/dependency-group
+import { ShortcutProvider } from '@wordpress/keyboard-shortcuts';
+
+/**
+ * Internal dependencies
+ */
+import { EditorWritingFlow } from './editor-writing-flow';
+import { registerBlocks } from './utils/register-blocks';
+
+registerBlocks();
+
+type RichTextEditorProps = {
+	blocks: BlockInstance[];
+	onChange: ( changes: BlockInstance[] ) => void;
+	entryId?: string;
+};
+
+export const RichTextEditor: React.VFC< RichTextEditorProps > = ( {
+	blocks,
+	onChange,
+} ) => {
+	const blocksRef = useRef( blocks );
+
+	const [ , setRefresh ] = useState( 0 );
+
+	useEffect( () => {
+		blocksRef.current = blocks;
+		// If there is a props change we need to update the ref and force re-render.
+		// Note: Because this component is memoized and because we don't re-render
+		// when this component initiates a change, a prop change won't force the re-render
+		// you'd expect. A change to the blocks must come from outside the editor.
+		setRefresh( ( r ) => r + 1 );
+	}, [ blocks ] );
+
+	// Use a combo of memoization and debounce to refresh every 200 milliseconds,
+	// ensuring that history is kept up to date for undo.
+	const debouncedRefresh = useCallback(
+		debounce( () => setRefresh( ( refresh ) => ( refresh += 1 ) ), 200 ),
+		[]
+	);
+
+	const debounceChange = debounce( ( updatedBlocks ) => {
+		onChange( updatedBlocks );
+		blocksRef.current = updatedBlocks;
+	}, 200 );
+
+	return (
+		<div className="woocommerce-rich-text-editor">
+			<SlotFillProvider>
+				<BlockEditorProvider
+					value={ blocksRef.current }
+					settings={ {
+						bodyPlaceholder: '',
+						hasFixedToolbar: true,
+						// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+						// @ts-ignore This property was recently added in the block editor data store.
+						__experimentalClearBlockSelection: false,
+					} }
+					onInput={ ( updatedBlocks ) => {
+						debounceChange( updatedBlocks );
+
+						debouncedRefresh();
```

Comment:
If we are gonna call `debouncedRefresh();` anytime `debounceChange( updatedBlocks );` is called I think we don't need a separete function for that; specially if that function containts a single statement line inside of it.
We can have something like 
```javascript
    const debounceChange = debounce( ( updatedBlocks ) => {
		onChange( updatedBlocks );
		blocksRef.current = updatedBlocks;
        setRefresh( ( refresh ) => refresh + 1 );
	}, 200 );
```

Reply:
I think you're right; it shouldn't be a problem to add the `debouncedRefresh` logic directly inside of here, but because we use this in a couple places (on `blocks` prop update as well), I moved this into a separate method `forceRerender` and moved the comment explaining how we use it above in https://github.com/woocommerce/woocommerce/pull/34865/commits/82f2e35b3e9f466ba4f08e69d087aaf487d0745f

This makes the intent more clear as to how and why we're using the `refresh` state.